chain/network/build.rs

```diff
@@ -0,0 +1,49 @@
+use sha2::Digest;
+
+fn content_prefix(content: &str) -> String {
+    format!(
+        "/// this file sha256 = {}\n",
+        base64::encode(&sha2::Sha256::digest(content.as_bytes()))
+    )
+}
+
+fn main() -> std::io::Result<()> {
+    // Generate code from proto, whenever proto files change.
+    // TODO: generalize to an arbitrary number of protos.
+
+    // TODO: this is disgusting, generated code shouldn't be checked into git.
+    let proto_path = "src/network_protocol/network.proto";
+    let generated_path = "src/network_protocol/generated/network.rs";
+
+    let want_hash = sha2::Sha256::digest(std::fs::read_to_string(proto_path)?.as_ref());
+    let mut want_prefix = String::new();
+    want_prefix += "/// This is an autogenerated file. DO NOT EDIT.\n";
+    want_prefix += &format!("/// proto file sha256 = {}\n", base64::encode(&want_hash));
+
+    let up_to_date = (|| {
+        let generated = std::fs::read_to_string(generated_path).ok()?;
+        let generated = generated.strip_prefix(&want_prefix)?;
+        let (prefix, content) = generated.split_once('\n')?;
+        if format!("{prefix}\n") != content_prefix(content) {
+            return None;
+        }
+        Some(())
+    })()
+    .is_some();
```

Comment:
Is there a reason you didn't use `starts_with`?

```suggestion
    let up_to_date = std::fs::read_to_string(generated_path).and_then(|s| {
        s.starts_with(&want_prefix)
    }).unwrap_or(false);
```

Reply:
It is more tricky than that. In addition to the hash of the proto file, I'm prepending the hash of the generated code to the file with the generated code ITSELF. So that if someone modifies the generated code by mistake, the compiler will should on them. To verify this hash, I need to first strip it from the file content, then compute the hash of the content remainder, and only then compare hashes.